docs/index.d.ts

```diff
@@ -13,7 +13,10 @@ export type RecognizedString = string | ArrayBuffer | Uint8Array | Int8Array | U
 /** A WebSocket connection that is valid from open to close event.
  * Read more about this in the user manual.
  */
-export interface WebSocket {
+export type WebSocket<UserData = {
```

Comment:
I think it would be better here to do UseerData extends and not = if it only can be an object

Reply:
Sorry, can you please give an example of what you mean?

Reply:
What about just making it ```WebSocket<UserData>```

Reply:
When you do 

```ts
UserData = {
    /** Arbitrary user data may be attached to this object. */
    [key: string]: any;
}
```

You can pass any type you want, but you have this one by default but you can pass `WebSocket<number>` if you want and getUserData() will return a number for typescript. So I said that if its always an objet we have to write

```ts
UserData extends {
    /** Arbitrary user data may be attached to this object. */
    [key: string]: any;
}
```

to force the TS user to only pass an "object" like type

Reply:
Sorry, @yovanoc, do you mean: `export type WebSocket<UserData extends {`?

If we change it to `UserData extends {` or `WebSocket<UserData>`, the code then requires the type argument which would break the current exported API.

Reply:
![image](https://user-images.githubusercontent.com/7302317/82629530-4d539100-9bf0-11ea-8507-f6e3c856d262.png)

You can always put `= {}` at the end

Reply:
![image](https://user-images.githubusercontent.com/7302317/82629585-770cb800-9bf0-11ea-902b-957e44230050.png)

Reply:
Just throw away that key value thing, it doesn't belong there it was added by someone else. It should just take a type, preferably object

Reply:
If getUserData was the only way to get data and it accept anything it would be perfect but that would cause a problem if we don't pass an object

![image](https://user-images.githubusercontent.com/7302317/82630267-812fb600-9bf2-11ea-9608-8126285a855a.png)

Reply:
Ahh, good catch @yovanoc.

So, if we want to keep the same API (not requiring the typing out all UserData), then we should do:

```ts
UserData extends {
    /** Arbitrary user data may be attached to this object. */
    [key: string]: any;
} = { [key: string]: any } 
```


I can't think of a library in typescript that requires types, are you sure you want me to remove default types?

Reply:
Problem with TypeScript is you lose the type information once it compiles to JS, so the template function ````ws<UserData>```` cannot properly call the constructors of UserData, because you don't know their type.

Regarding extends, isn't there some "extends Object" kind of thing?

Reply:
The only way to make this interface work would be to wrap the entire API in a TypeScript wrapper that takes the type of UserData and makes sure to call the constructor/destructor of UserData before/after emitting open/close events.

And that sucks.